### PR TITLE
Adding checkpointing infrastructure for easier space recovery.

### PR DESCRIPTION
**Pull request by mmcdermott**
Checkpoints are taken every save.

### CODE REVIEW

**Comment by mmcdermott** [comment]
We can also add two checkpointing classes, one for 'quick checkpoints' that are taken every save and one for long lasting checkpoints that are taken every hour or so, with different queue conditions on each, but that felt overly complex for a first pass.

File: app/models/space.rb [hunk 1]
```diff
@@ -126,6 +128,20 @@ def copy(user)
...
+
+  def needs_checkpoint?
+    checkpoint = checkpoints.order("created_at").last
+    return false unless checkpoint && graph
+    15.minutes.ago >= checkpoint.created_at
+  end
+
+  def take_checkpoint
+    checkpoints.create name: name, description: description, graph: graph
+    if checkpoints.count > 3000 # 3000 at 15 minute intervals in 1 month
```
Comment (OAGr) [inline app/models/space.rb]: What?  I doubt they will be continuously editing for 1 month.  For now, how about 30?  Can scale after

File: app/models/space.rb [hunk 2]
```diff
@@ -126,6 +128,20 @@ def copy(user)
     return space
   end
 
+
+  def needs_checkpoint?
+    checkpoint = checkpoints.order("created_at").last
```
Comment (OAGr) [inline app/models/space.rb]: Single quotes

**Comment by OAGr** [comment]
I think a worker or delayed system, and images are the next two main features to add.  Let's try this out for now though and see how it works out.

**Pull request merged**